pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go

```diff
@@ -226,105 +227,99 @@ func (c *ScalingJob) addFailedExecutionHistory(
 	cronFHPA *autoscalingv1alpha1.CronFederatedHPA, errMsg string) error {
 	_, nextExecutionTime := c.scheduler.NextRun()
 
-	// Add success history record, return false if there is no such rule
-	addFailedHistoryFunc := func() bool {
-		exists := false
-		for index, rule := range cronFHPA.Status.ExecutionHistories {
-			if rule.RuleName != c.rule.Name {
-				continue
-			}
-			failedExecution := autoscalingv1alpha1.FailedExecution{
-				ScheduleTime:  rule.NextExecutionTime,
-				ExecutionTime: &metav1.Time{Time: time.Now()},
-				Message:       errMsg,
-			}
-			historyLimits := helper.GetCronFederatedHPAFailedHistoryLimits(c.rule)
-			if len(rule.FailedExecutions) > historyLimits-1 {
-				rule.FailedExecutions = rule.FailedExecutions[:historyLimits-1]
-			}
-			cronFHPA.Status.ExecutionHistories[index].FailedExecutions =
-				append([]autoscalingv1alpha1.FailedExecution{failedExecution}, rule.FailedExecutions...)
-			cronFHPA.Status.ExecutionHistories[index].NextExecutionTime = &metav1.Time{Time: nextExecutionTime}
-			exists = true
-			break
+	// Add failed history record
+	addFailedHistoryFunc := func(index int) {
+		failedExecution := autoscalingv1alpha1.FailedExecution{
+			ScheduleTime:  cronFHPA.Status.ExecutionHistories[index].NextExecutionTime,
+			ExecutionTime: &metav1.Time{Time: time.Now()},
+			Message:       errMsg,
 		}
-
-		return exists
+		historyLimits := helper.GetCronFederatedHPAFailedHistoryLimits(c.rule)
+		if len(cronFHPA.Status.ExecutionHistories[index].FailedExecutions) > historyLimits-1 {
+			cronFHPA.Status.ExecutionHistories[index].FailedExecutions = cronFHPA.Status.ExecutionHistories[index].FailedExecutions[:historyLimits-1]
+		}
+		cronFHPA.Status.ExecutionHistories[index].FailedExecutions =
+			append([]autoscalingv1alpha1.FailedExecution{failedExecution}, cronFHPA.Status.ExecutionHistories[index].FailedExecutions...)
+		cronFHPA.Status.ExecutionHistories[index].NextExecutionTime = &metav1.Time{Time: nextExecutionTime}
 	}
 
-	return retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
-		// If this history not exist, it means the rule is suspended or deleted, so just ignore it.
-		if exists := addFailedHistoryFunc(); !exists {
+	var operationResult controllerutil.OperationResult
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+		operationResult, err = helper.UpdateStatus(context.Background(), c.client, cronFHPA, func() error {
+			index := c.findExecutionHistory(cronFHPA.Status.ExecutionHistories)
+			if index < 0 {
+				// The failed history does not exist, it means the rule deleted, so just ignore it.
+				return nil
```

Comment:
The issue here is that `return` doesn't skip the updation operation, it just skips mutating, which means `UpdateStatus` probably still performs the updation.

Reply:
do you have any further comments?

Reply:
The cheap way is probably just to check rule existence before `UpdateStatus`, even before `RetryOnConflict`. something like:
```diff
diff --git a/pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go b/pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go 
index 1a250a4d9..899bed058 100755 
--- a/pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go 
+++ b/pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go
@@ -243,14 +243,15 @@ func (c *ScalingJob) addFailedExecutionHistory( 
                cronFHPA.Status.ExecutionHistories[index].NextExecutionTime = &metav1.Time{Time: nextExecutionTime}
        }
 
+       index := c.findExecutionHistory(cronFHPA.Status.ExecutionHistories) 
+       if index < 0 { 
+               // The failed history does not exist, it means the rule deleted, so just ignore it.
+               return nil 
+       } 
+ 
        var operationResult controllerutil.OperationResult
        if err := retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
                operationResult, err = helper.UpdateStatus(context.Background(), c.client, cronFHPA, func() error {
-                       index := c.findExecutionHistory(cronFHPA.Status.ExecutionHistories)
-                       if index < 0 {
-                               // The failed history does not exist, it means the rule deleted, so just ignore it.
-                               return nil
-                       }
                        addFailedHistoryFunc(index)
                        return nil
                })
```

Reply:
I think the CronFederatedHPA can have methods like
```golang
func (c *CronFederatedHPA) ExecutionHistoryForRuleExist(ruleName string) bool {
	
}

func (c *CronFederatedHPA) AppendFailedExecutionHistoryForRule(failedHistory FailedExecution) {
	
}
```
But that's beyond the scope of this PR. We can do it after this PR.